### PR TITLE
Feature #753

### PR DESCRIPTION
**Pull request by PGijsbers**
Looking for feedback. For the example I still need to finalize how the predictions are generated/formatted to make it clear and not distract from the overall example.

### CODE REVIEW

File: examples/30_extended/custom_flow_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+#
+# Note: flow 15275 is not actually the right flow on the test server,
+# but that does not matter for this demonstration.
+
+autosklearn_flow = openml.flows.get_flow(15275)  # auto-sklearn 0.5.1
+subflow = dict(components=OrderedDict(automl_tool=autosklearn_flow),)
+
+####################################################################################################
+# With all parameters of the flow defined, we can now initialize the OpenMLFlow and publish.
+# Explicitly set the model of the flow to `None`, because we provided all the details already!
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:86]: I would state this differently. The model is set to `None`, because there is no model to access and parse here, right?

File: examples/30_extended/custom_flow_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    for index in task.get_train_test_split_indices(fold, repeat)[1]
+]
+
+# random class probabilities (Iris has 150 samples and 3 classes):
+r = np.random.rand(150 * n_repeats, 3)
+# scale the random values so that the probabilities of each sample sum to 1:
+y_proba = r / r.sum(axis=1).reshape(-1, 1)
+y_pred = y_proba.argmax(axis=1)
+class_map = dict(zip(range(3), task.class_labels))
+y_true = ["Iris-setosa"] * 50 + ["Iris-versicolor"] * 50 + ["Iris-virginica"] * 50
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:146]: I think it would be better for the example if you retrieved these from the dataset.

File: examples/30_extended/custom_flow_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        index=index,
+        prediction=class_map[yp],
+        truth=y,
+        proba={c: pb for (c, pb) in zip(task.class_labels, proba)},
+    )
+    ps.append(p)
+
+####################################################################################################
+# Finally we can create the OpenMLRun object and upload.
+# We use the "setup string" because the used flow was a script.
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:169]: Can you rephrase this, it took me a while to understand what you mean here. Maybe write "We use the argument ``setup_string`` because the used flow was a script."

File: examples/30_extended/custom_flow_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        task=task,
+        repeat=repeat,
+        fold=fold,
+        sample=sample,
+        index=index,
+        prediction=class_map[yp],
+        truth=y,
+        proba={c: pb for (c, pb) in zip(task.class_labels, proba)},
+    )
+    ps.append(p)
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:165]: This is not used, right? Was this the variable that you mentioned we shouldn't have a look at yesterday?

Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:165]: I don't recall what I mentioned yesterday, but yes this was left over from development/checking the function runs fine. But yes I will remove it :)

**Comment by PGijsbers** [comment]
Thanks for the feedback :) will process after the meeting.

File: examples/30_extended/custom_flow_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
+"""
+================================
+Creating and Using a Custom Flow
+================================
+
+The most convenient way to create a flow for your machine learning workflow is to generate it
+automatically as described in <>. However, there are scenarios where this is not possible, such
```
Comment (Bilgecelik) [inline examples/30_extended/custom_flow_tutorial.py:7]: Maybe one sentence before this to explain why they need flows at the first place?

File: examples/30_extended/custom_flow_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+####################################################################################################
+# 1. Defining the flow
+# ====================
+# The first step is to define all the hyperparameters of your flow.
+# Check ... for the descriptions of each variable.
+# Note that `external version` and `name` together should uniquely identify a flow.
+#
+# The AutoML Benchmark runs AutoML systems across a range of tasks.
+# We can not use the flows of the AutoML systems directly, as the benchmark adds performs
+# preprocessing as required.
```
Comment (Bilgecelik) [inline examples/30_extended/custom_flow_tutorial.py:39]: The second part is not clear to me

File: examples/30_extended/custom_flow_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        cores=OrderedDict(description="number of available cores", data_type="int"),
+        memory=OrderedDict(description="memory in gigabytes", data_type="int"),
+        time=OrderedDict(description="time in minutes", data_type="int"),
+    ),
+)
+
+####################################################################################################
+# It is possible for flows to contain subflows. In this example, the auto-sklearn flow is a
+# subflow, this means that the subflow is entirely executed as part of this flow.
+# Using this modularity also allows your runs to specify which hyperparameters of the
```
Comment (Bilgecelik) [inline examples/30_extended/custom_flow_tutorial.py:75]: Is it also possible to define a custom flow without defining it as a subflow? So we need subflows when we want to do multiple runs with different hyperparameter settings for each subflow?

File: examples/30_extended/custom_flow_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    OrderedDict([("oml:name", "time"), ("oml:value", 120), ("oml:component", flow_id)]),
+]
+
+task_id = 1408  # Iris Task
+task = openml.tasks.get_task(task_id)
+dataset_id = task.get_dataset().dataset_id
+
+
+####################################################################################################
+# The last bit of information for the run we need are the predicted values.
```
Comment (Bilgecelik) [inline examples/30_extended/custom_flow_tutorial.py:116]: I think this part may use some more explanation/categorization but not sure how=) Maybe explain what is different for custom flow here, and why.

File: examples/30_extended/custom_flow_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    ),
+    external_version="amlb==0.9",
+    language="English",
+    tags=["amlb", "benchmark", "study_218"],
+    dependencies="amlb==0.9",
+)
+
+####################################################################################################
+# Next we define the flow hyperparameters. We define their name and default value in `parameters`,
+# and provide meta-data for each parameter through `parameters_meta_info`.
```
[the remark this answers is not in the document]
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:60]: Yes, though the function arguments make it a little confusing.

File: examples/30_extended/custom_flow_tutorial.py (hunk 1 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:86]: In the benchmark case, there is not. However in the general case there could be a python object model, but with no extension for automatically populating e.g. `parameters` fields.

File: examples/30_extended/custom_flow_tutorial.py (hunk 6 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:39]: Changes as 

```
# OpenML stores Flows for each AutoML system. However, the AutoML benchmark adds
# preprocessing to the flow, so should be described in a new flow.
```

is that clear? Or could you try to identify the part that raises confusion?

File: examples/30_extended/custom_flow_tutorial.py (hunk 7 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:75]: It is possible to have custom flows with no subflows. Custom flows (such as this one) can be defined with their own hyperparameters (such as time or memory for this one).

File: examples/30_extended/custom_flow_tutorial.py (hunk 8 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:116]: There is nothing inherently different for custom flows here. But when you run flows through extensions, (i.e. use `run_flow_on_task`) all of this formatting is done for you automatically.

File: examples/30_extended/custom_flow_tutorial.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mfeurer) [reply examples/30_extended/custom_flow_tutorial.py]: hm, yes :( Maybe you should briefly mention this in the example, too?

File: examples/30_extended/custom_flow_tutorial.py (hunk 1 above)
Comment (mfeurer) [reply examples/30_extended/custom_flow_tutorial.py:87]: Yes, I see. But it sounds like you don't need to provide a model if all information is already parsed. I think it would be better if you could briefly mention the possible cases and which one you chose in this example.

File: examples/30_extended/custom_flow_tutorial.py [hunk 10]
```diff
@@ -142,31 +150,31 @@
...
         truth=y,
         proba={c: pb for (c, pb) in zip(task.class_labels, proba)},
     )
-    ps.append(p)
+    predictions.append(prediction)
 
 ####################################################################################################
 # Finally we can create the OpenMLRun object and upload.
-# We use the "setup string" because the used flow was a script.
+# We use the argument setup_string because the used flow was a script."
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:177]: I think you accidentally added a trailing quotation mark here?

File: examples/30_extended/custom_flow_tutorial.py (hunk 6 above)
Comment (Bilgecelik) [reply examples/30_extended/custom_flow_tutorial.py]: No it is clear now.

File: examples/30_extended/custom_flow_tutorial.py (hunk 7 above)
Comment (Bilgecelik) [reply examples/30_extended/custom_flow_tutorial.py:76]: I understand, i am still a bit unsure of flow / custom flow / subflow differences. Maybe also additional value of using subflows instead of multiple custom flows, but that can be just me=)

File: examples/30_extended/custom_flow_tutorial.py (hunk 8 above)
Comment (Bilgecelik) [reply examples/30_extended/custom_flow_tutorial.py:124]: okay then it is fine

File: examples/30_extended/custom_flow_tutorial.py (hunk 7 above)
Comment (Bilgecelik) [reply examples/30_extended/custom_flow_tutorial.py]: I think it is quite clear now.

**Comment by PGijsbers** [comment]
@mfeurer I don't see any reduced coverage, (where) am I reading the codecov report wrong?

**Comment by mfeurer** [comment]
I think it complains that not enough of the diff is tested. 76% of your diff is tested, while codecov expects 87.65% of the diff being tested. I guess the 87.65% is our current test coverage.

**Comment by PGijsbers** [comment]
Ah, I was looking at the wrong tab 😓  looks like I'll have to add some tests that check the error cases / learning curve tasks. Thanks!

File: examples/30_extended/custom_flow_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    external_version="amlb==0.9",
+    language="English",
+    tags=["amlb", "benchmark", "study_218"],
+    dependencies="amlb==0.9",
+)
+
+####################################################################################################
+# Next we define the flow hyperparameters. We define their name and default value in `parameters`,
+# and provide meta-data for each hyperparameter through `parameters_meta_info`.
+# Note that eventhough the argument name is `parameters` they describe the hyperparameters.
```
Comment (Neeratyoy) [inline examples/30_extended/custom_flow_tutorial.py:61]: "eventhough" -> "even though"

File: examples/30_extended/custom_flow_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+####################################################################################################
+# It is possible to build a flow which uses other flows.
+# For example, the Random Forest Classifier is a flow, but you could also construct a flow
+# which uses a Random Forest Classifier in a ML pipeline. When constructing the pipeline flow,
+# you can use the Random Forest Classifier flow as a *subflow*. It allows for
+# all hyperparameters of the Random Classifier Flow to also be specified in your pipeline flow.
+#
+# In this example, the auto-sklearn flow is a subflow: the auto-sklearn flow is entirely executed as part of this flow.
+# This allows people to specify auto-sklearn hyperparameters used in this flow.
+# Using a subflow is not required.
```
Comment (Neeratyoy) [inline examples/30_extended/custom_flow_tutorial.py:82]: I think I understood the earlier part about an existing flow included as a *subflow* in the new custom flow. However, I am a little confused with these lines here where it is mentioned that the auto-sklearn flow is executed as a subflow, but then what does "using a subflow is not required" mean here?

File: examples/30_extended/custom_flow_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+#
+# Note: flow 15275 is not actually the right flow on the test server,
+# but that does not matter for this demonstration.
+
+autosklearn_flow = openml.flows.get_flow(15275)  # auto-sklearn 0.5.1
+subflow = dict(components=OrderedDict(automl_tool=autosklearn_flow),)
```
Comment (Neeratyoy) [inline examples/30_extended/custom_flow_tutorial.py:88]: The dict key of *automl_tool* is a registered/expected key or the user is free to rename it while creating their flow?
If it is the former, could we then mention or link here how the user could look up other such keys they may require to be specified?

File: examples/30_extended/custom_flow_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+# - p_1, ..., p_c: for each class the predicted probability of the sample
+#   belonging to that class. (no elements for regression tasks)
+#   Make sure the order of these elements follows the order of `task.class_labels`.
+# - the predicted class/value for the sample
+# - the true class/value for the sample
+#
+# When using openml-python extensions (such as through `run_model_on_task`),
+# all of this formatting is automatic.
+# Unfortunately we can not automate this procedure for custom flows,
+# which means a little additional effort is required.
```
Comment (Neeratyoy) [inline examples/30_extended/custom_flow_tutorial.py:145]: For me personally, the minor gap in understanding that exists is the behaviour of functions available in the OpenML API. Given that the example so far creates a *custom* flow, pulls a task and a dataset and then jumps to how the predictions should be formatted for a legitimate OpenML run format, leaves me wondering whether I can use any OpenML function to reach till the prediction formatting stage. 
As it appears, one would have to pull the dataset using *task* API such as `get_X_and_y`, and explicitly call fit(), predict() or the equivalent, and then format the output predictions as you mention later in the example.

If I have understood it correctly, then I think that having at least comments talking briefly about the next imminent steps to go from flow, task, data to predictions might be helpful. Though on second thought, what I said feels like quite the obvious next steps when dealing with such a flow, maybe being explicit might help the new OpenML users. 

To summarise, the fact that `run_model_on_task` may not work on a custom flow without a registered extension could be more explicit, if that is indeed what is supposed to happen.

File: examples/30_extended/custom_flow_tutorial.py (hunk 12 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:82]: It is just to say that not every flow has to contain a subflow, but this one does because it's useful. What do you think about

> In general, using a subflow is not required.

instead?

File: examples/30_extended/custom_flow_tutorial.py (hunk 13 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:88]: It's essentially the name I give to the subflow (see https://openml.org/f/15509). You're free to pick the name. For example in the scikit-learn pipeline of [flow 8817](https://www.openml.org/f/8817) you find components like `ColumnTransformer` and `SVC` instead.

File: examples/30_extended/custom_flow_tutorial.py (hunk 14 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:145]: > As it appears, one would have to pull the dataset using task API such as get_X_and_y, and explicitly call fit(), predict() or the equivalent, and then format the output predictions as you mention later in the example.

Correct, as custom flows are used when OpenML can't interface directly with the machine learning flow.

> If I have understood it correctly, then I think that having at least comments talking briefly about the next imminent steps to go from flow, task, data to predictions might be helpful.

I am not sure what you are missing here, all of this is documented step by step in this tutorial? Exactly how to feed data to the model and produce predictions is dependent on the framework you work with. In our case we just generate predictions instead, but in general we don't know the interface which is why it's not available as an extension in the first place.

> To summarise, the fact that run_model_on_task may not work on a custom flow without a registered extension could be more explicit, if that is indeed what is supposed to happen.

In which way do you feel the opening paragraphs do not address this? (I did just notice I forgot to put in the reference link, so perhaps that's why)

File: examples/30_extended/custom_flow_tutorial.py (hunk 13 above)
Comment (Neeratyoy) [reply examples/30_extended/custom_flow_tutorial.py:89]: aah okay, got it
all good then

File: examples/30_extended/custom_flow_tutorial.py (hunk 12 above)
Comment (Neeratyoy) [reply examples/30_extended/custom_flow_tutorial.py]: Yeah, sounds much clearer for me personally

File: examples/30_extended/custom_flow_tutorial.py (hunk 14 above)
Comment (Neeratyoy) [reply examples/30_extended/custom_flow_tutorial.py:146]: yes, the link makes it clear :)

File: openml/runs/functions.py [hunk 15]
```diff
@@ -1073,4 +1075,7 @@ def format_prediction(
                 sample = 0
         probabilities = [proba[c] for c in task.class_labels]
         return [repeat, fold, sample, index, *probabilities, truth, prediction]
-    return [repeat, fold, index, truth, prediction]
+    elif isinstance(task, OpenMLRegressionTask):
+        return [repeat, fold, index, truth, prediction]
+    else:
+        raise TypeError(f"Formatting for {type(task)} is not supported.")
```
Comment (mfeurer) [inline openml/runs/functions.py:1081]: Would a `NotImplementedError` be more logical here?

File: tests/test_runs/test_run_functions.py [hunk 16]
```diff
@@ -1336,3 +1333,44 @@ def test_run_flow_on_task_downloaded_flow(self):
...
+        openml.config.server = self.production_server
+        clustering = openml.tasks.get_task(126033, download_data=False)
+        ignored_input = [0] * 5
+        with self.assertRaises(TypeError):
+            format_prediction(clustering, *ignored_input)
+
+    def test_format_prediction_classification_no_probabilities(self):
+        classification = openml.tasks.get_task(self.TEST_SERVER_TASK_SIMPLE[0], download_data=False)
+        ignored_input = [0] * 5
+        with self.assertRaises(ValueError):
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1348]: I'd actually prefer to have `self.assertRaisesRegex` so that one sees what exact error is expected. Could you please change that?

File: tests/test_runs/test_run_functions.py [hunk 17]
```diff
@@ -1336,3 +1333,44 @@ def test_run_flow_on_task_downloaded_flow(self):
...
+            format_prediction(classification, *ignored_input, proba=None)
+
+    def test_format_prediction_classification_incomplete_probabilities(self):
+        classification = openml.tasks.get_task(self.TEST_SERVER_TASK_SIMPLE[0], download_data=False)
+        ignored_input = [0] * 5
+        incomplete_probabilities = {c: 0.2 for c in classification.class_labels[1:]}
+        with self.assertRaises(ValueError):
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1355]: Same here.

File: tests/test_runs/test_run_functions.py [hunk 18]
```diff
@@ -1336,3 +1333,44 @@ def test_run_flow_on_task_downloaded_flow(self):
...
+            format_prediction(classification, *ignored_input, proba=incomplete_probabilities)
+
+    def test_format_prediction_task_without_classlabels_set(self):
+        classification = openml.tasks.get_task(self.TEST_SERVER_TASK_SIMPLE[0], download_data=False)
+        classification.class_labels = None
+        ignored_input = [0] * 5
+        with self.assertRaises(ValueError):
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1362]: and here

File: tests/test_runs/test_run_functions.py [hunk 19]
```diff
@@ -1336,3 +1333,44 @@ def test_run_flow_on_task_downloaded_flow(self):
...
+            format_prediction(classification, *ignored_input, proba={})
+
+    def test_format_prediction_task_learning_curve_sample_not_set(self):
+        learning_curve = openml.tasks.get_task(801, download_data=False)
+        probabilities = {c: 0.2 for c in learning_curve.class_labels}
+        ignored_input = [0] * 5
+        with self.assertRaises(ValueError):
```
Comment (mfeurer) [inline tests/test_runs/test_run_functions.py:1369]: and here as well

File: examples/30_extended/custom_flow_tutorial.py [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+Creating and Using a Custom Flow
+================================
+
+The most convenient way to create a flow for your machine learning workflow is to generate it
+automatically as described in the `Obtain Flow IDs <https://openml.github.io/openml-python/master/examples/30_extended/flow_id_tutorial.html#sphx-glr-examples-30-extended-flow-id-tutorial-py>`_ tutorial.  # noqa E501
+However, there are scenarios where this is not possible, such
+as when the flow uses a framework without an extension or when the flow is described by a script.
+
+In those cases you can still create a custom flow by following the steps of this tutorial.
+As an example we will use the flows generated for the AutoML Benchmark (...),
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:12]: I believe there is a link missing here (and it shouldn't be `(...)`)

File: examples/30_extended/custom_flow_tutorial.py [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+from openml import OpenMLClassificationTask
+from openml.runs.functions import format_prediction
+
+openml.config.start_using_configuration_for_example()
+
+####################################################################################################
+# 1. Defining the flow
+# ====================
+# The first step is to define all the hyperparameters of your flow.
+# Check ... for the descriptions of each variable.
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:35]: Is there a link missing here, too?

File: examples/30_extended/custom_flow_tutorial.py [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+
+####################################################################################################
+# It is possible to build a flow which uses other flows.
+# For example, the Random Forest Classifier is a flow, but you could also construct a flow
+# which uses a Random Forest Classifier in a ML pipeline. When constructing the pipeline flow,
+# you can use the Random Forest Classifier flow as a *subflow*. It allows for
+# all hyperparameters of the Random Classifier Flow to also be specified in your pipeline flow.
+#
+# In this example, the auto-sklearn flow is a subflow: the auto-sklearn flow is entirely executed as part of this flow.
+# This allows people to specify auto-sklearn hyperparameters used in this flow.
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:82]: I am wondering if this is necessary for the example. Isn't the Auto-sklearn flow itself sufficient? And wouldn't it be better to create a custom flow from scratch to explain the idea of creating flows as a user?

File: examples/30_extended/custom_flow_tutorial.py [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+    flow_id=flow_id,
+    dataset_id=dataset_id,
+    parameter_settings=parameters,
+    setup_string=benchmark_command,
+    data_content=predictions,
+    tags=["study_218"],
+    description_text="Run generated by the Custom Flow tutorial.",
+)
+my_run.publish()
+print("run created:", my_run.run_id)
```
Comment (mfeurer) [inline examples/30_extended/custom_flow_tutorial.py:203]: I think you should use the "stop using the test server command" in the end of the example to go back to the default setting.

**Review by mfeurer — COMMENTED** [review]
Hey Pieter, I'm afraid I have some more questions on this example (and a few minor change requests)

File: examples/30_extended/custom_flow_tutorial.py (hunk 22 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:82]: In the context of the example, yes we could use a flow without subflow. It would make the example easier, though then there is no documentation on how subflows should be referenced. If you think it is better to leave that out, I can simplify the example.

The AutoML benchmark flows are structured this way (with a subflow), because the benchmark actually performs operations on the data: it will encode categorical data and impute missing values if the framework is not able to work directly with missing values or string data.

File: tests/test_runs/test_run_functions.py (hunk 16 above)
Comment (PGijsbers) [reply tests/test_runs/test_run_functions.py:1348]: I've actually been meaning to ask. Why do we use the `unittest` module for some things, and `pytest` for others?
For instance, instead of `self.assertRaisesRegex`  you would use `pytest.raises(ValueError, match=r'regex string')`.

File: examples/30_extended/custom_flow_tutorial.py (hunk 22 above)
Comment (mfeurer) [reply examples/30_extended/custom_flow_tutorial.py:82]: > If you think it is better to leave that out, I can simplify the example.

Ideally we'd have both, not sure how we can easily get this done without having two examples in the end

File: tests/test_runs/test_run_functions.py (hunk 16 above)
Comment (mfeurer) [reply tests/test_runs/test_run_functions.py:1348]: Historical reasons ;)

I think one could actually only use `assert a == b` and not use the `unittest` or `pytest` module any more, see https://docs.pytest.org/en/stable/assert.html

Comment (PGijsbers) [reply tests/test_runs/test_run_functions.py:1348]: Yes, but just `assert` does not work for checking exceptions. The page you linked describes the `pytest.raises(ValueError, match=r'regex string')` call I referred too as well.

Comment (PGijsbers) [reply tests/test_runs/test_run_functions.py:1348]: I'll use the `unittest` `self.assertRaisesRegex` for consistency.

File: openml/runs/functions.py (hunk 15 above)
Comment (PGijsbers) [reply openml/runs/functions.py:1081]: Sure, updated.

File: examples/30_extended/custom_flow_tutorial.py (hunk 22 above)
Comment (PGijsbers) [reply examples/30_extended/custom_flow_tutorial.py:82]: Would it be acceptable to extend the example with a flow without subflow later? This example at least covers that implicitly anyway. I could add it when I get back from vacation or during the hackathon.

Comment (mfeurer) [reply examples/30_extended/custom_flow_tutorial.py:82]: > Would it be acceptable to extend the example with a flow without subflow later?

Yes, I think so.

**Review by mfeurer — APPROVED** [review]

**Comment by PGijsbers** [comment]
I did a rebase on develop, locally it looked like it worked fine. I thought the old commits should've simply been replaced by the ones patched onto the develop head? (i.e. same code diff, different commit id)

**Comment by mfeurer** [comment]
> I thought the old commits should've simply been replaced by the ones patched onto the develop head?

Yeah, I don't know how to do this probably either, I'm usually rebasing to avoid such hassle.

> I can try to fix this, but I guess we are squashing anyway so it doesn't matter much?

No, shouldn't matter.

Shall I do a final review and then merge?

**Comment by PGijsbers** [comment]
> Yeah, I don't know how to do this probably either, I'm usually rebasing to avoid such hassle.

I normally merge because it's less of a hassle. After the failed rebase I actually found the general advise not to rebase if the work already lives on remote.

> Shall I do a final review and then merge?

That would be greatly appreciated!
